src/ape_cache/query.py

```diff
@@ -0,0 +1,337 @@
+        )
+
+    @perform_query_clause.register
+    def perform_block_clause(self, query: BlockQuery) -> TextClause:
+        return text(
+            """
+    SELECT :columns
+    FROM blocks
+    WHERE blocks.number >= :start_block
+    AND blocks.number <= :stop_block
+    AND blocks.number % :step = 0
+        """
+        ).bindparams(
+            columns=",".join(query.columns),
+            start_block=query.start_block,
+            stop_block=query.stop_block,
+            step=query.step,
+        )
+
+    @perform_query_clause.register
+    def perform_transaction_clause(self, query: BlockTransactionQuery) -> TextClause:
+        return text(
+            """
+    SELECT :columns
+    FROM transactions
+    WHERE transactions.block_hash = :block_id
+        """
+        ).bindparams(columns=",".join(query.columns), block_id=query.block_id)
+
+    @perform_query_clause.register
+    def perform_contract_event_clause(self, query: ContractEventQuery) -> TextClause:
+        return text(
+            """
+    SELECT :columns
+    FROM contract_events
+    WHERE contract_events.block_number >= :start_block
+    AND contract_events.block_number <= :stop_block
+    AND contract_events.block_number % :step = 0
+        """
+        ).bindparams(
+            columns=",".join(query.columns),
+            start_block=query.start_block,
+            stop_block=query.stop_block,
+            step=query.step,
+        )
+
+    def perform_query(self, query: QueryType) -> Optional[pd.DataFrame]:  # type: ignore
+        try:
+            with self.database_connection as conn:
+                result = conn.execute(self.perform_query_clause(query))
+
+                if not result:
+                    # NOTE: Should be unreachable if estimated correctly
+                    raise QueryEngineError(f"Could not perform query:\n{query}")
+
+                # TODO: Fix this, should return an iterator
+                return pd.DataFrame(columns=query.columns, data=result.fetchall())
+
+        except QueryEngineError as err:
+            logger.error(f"Database not initiated: {str(err)}")
+
+    @singledispatchmethod
+    def cache_update_clause(self, query: QueryType) -> Optional[Insert]:
+        pass  # Can't cache this query
+
+    @cache_update_clause.register
+    def cache_update_block_clause(self, query: BlockQuery) -> Optional[Insert]:
+        return insert(Blocks)  # type: ignore
+
+    @cache_update_clause.register
+    def cache_update_block_txns_clause(self, query: BlockTransactionQuery) -> Optional[Insert]:
+        return insert(Transactions)  # type: ignore
+
+    @cache_update_clause.register
+    def cache_update_events_clause(self, query: ContractEventQuery) -> Optional[Insert]:
+        return insert(ContractEvents)  # type: ignore
+
+    @singledispatchmethod
+    def get_cache_data(
+        self, query: QueryType, result: Iterator[BaseInterfaceModel]
+    ) -> Optional[List[Dict[str, Any]]]:
+        raise QueryEngineError(
```

Comment:
We could abstract this into another class since it is used in multiple spots:

```
class QueryTypeError(QueryEngineError):
   def __init__():
      super().__init__(
        "Not a compatible QueryType. For more details see our docs"
        "https://docs.apeworx.io/ape/stable/methoddocs/exceptions.html#ape.exceptions.QueryEngineError"
      )   
```

Reply:
Should I create an issue for this?

Reply:
No, I think this may be a good suggestion for here. You can add to `src/ape_cache/exceptions.py`